src/sentry/eventstream/kafka/postprocessworker.py

```diff
@@ -0,0 +1,134 @@
+import logging
+import random
+from concurrent.futures import Future, ThreadPoolExecutor, as_completed
+from contextlib import contextmanager
+from typing import Any, Generator, Mapping, Optional, Sequence
+
+from sentry import options
+from sentry.eventstream.kafka.protocol import (
+    get_task_kwargs_for_message,
+    get_task_kwargs_for_message_from_headers,
+)
+from sentry.tasks.post_process import post_process_group
+from sentry.utils import metrics
+from sentry.utils.batching_kafka_consumer import AbstractBatchWorker
+from sentry.utils.cache import cache_key_for_event
+
+logger = logging.getLogger(__name__)
+
+Message = Any
+_DURATION_METRIC = "eventstream.duration"
+_CONCURRENCY_OPTION = "post-process-forwarder:concurrency"
+
+
+@contextmanager
+def _sampled_eventstream_timer(instance: str) -> Generator[None, None, None]:
+    record_metric = random.random() < 0.1
+    if record_metric is True:
+        with metrics.timer(_DURATION_METRIC, instance=instance):
+            yield
+    else:
+        yield
+
+
+def _get_task_kwargs(message: Message) -> Optional[Mapping[str, Any]]:
+    use_kafka_headers = options.get("post-process-forwarder:kafka-headers")
+
+    if use_kafka_headers:
+        try:
+            with _sampled_eventstream_timer(instance="get_task_kwargs_for_message_from_headers"):
+                return get_task_kwargs_for_message_from_headers(message.headers())
+        except Exception as error:
+            logger.error("Could not forward message: %s", error, exc_info=True)
+            with metrics.timer(_DURATION_METRIC, instance="get_task_kwargs_for_message"):
+                return get_task_kwargs_for_message(message.value())
+    else:
+        with metrics.timer(_DURATION_METRIC, instance="get_task_kwargs_for_message"):
+            return get_task_kwargs_for_message(message.value())
```

Comment:
This has been taken from https://github.com/getsentry/sentry/blob/master/src/sentry/eventstream/kafka/backend.py#L281-L306

With the minor change that dispatching of the task happens in a different function